### PR TITLE
Improved creation of referenced entities

### PR DESCRIPTION
**Pull request by mtodor**

### CODE REVIEW

File: src/ContentCreator.php [hunk 1]
```diff
@@ -615,6 +669,34 @@ protected function getEntityBundleInstances($entity_type) {
...
+   */
+  protected function getEntityBundleInstancesForField($entity_type, $bundle_type, $field_name) {
+    if (isset($this->cacheBundleInstancesForField[$entity_type][$bundle_type][$field_name])) {
+      return $this->cacheBundleInstancesForField[$entity_type][$bundle_type][$field_name];
+    }
+
+    $field_target_bundle_info = $this->config[$entity_type]['_bundles'][$bundle_type]['_fields'][$field_name]['_bundle_info'];
+    $this->cacheBundleInstancesForField[$entity_type][$bundle_type][$field_name] = array_filter($this->getEntityBundleInstances($field_target_bundle_info['target_type']), function ($target_entity_instances, $target_bundle_type) use ($field_target_bundle_info) {
+      return empty($field_target_bundle_info['target_bundles']) || isset($field_target_bundle_info['target_bundles'][$target_bundle_type]);
+    }, ARRAY_FILTER_USE_BOTH);
```
Comment (dbosen) [inline src/ContentCreator.php]: `$target_entity_instances` is not used, why use ARRAY_FILTER_USE_BOTH then?

File: src/ContentCreator.php [hunk 2]
```diff
@@ -547,38 +602,37 @@ protected function getBundleReferencedFieldDefinitions($entity_type, $bundle_typ
...
-        }
-
         $num_of_instances = $this->getRandomWeighted($histogram);
 
-        // In case of recursive nesting, we are not going to add entities.
-        if ($this->createBundle($target_entity_type, $target_entity_bundle, $num_of_instances, $entity_id, $entity_type, $field_name) === 0) {
-          continue;
+        $target_entity_bundles = [];
+        for ($instance_index = $num_of_instances; $instance_index >= 0; $instance_index--) {
+          $target_entity_bundles[] = $this->getRandomWeighted($this->getEntityBundleInstancesForField($entity_type, $bundle_type, $field_name));
```
Comment (dbosen) [inline src/ContentCreator.php]: `getEntityBundleInstancesForField` can return empty array, that will create empty values for `$target_entity_bundles`. If such an `$target_entity_bundles` array is created, the foreach below will have empty `$target_entity_bundle` and `$this->createBundleForReferenceField` will throw "Undefined index:" errors.
I have created a report, which leads to that error. I will send it to you, so you can investigate this.

**Review by dbosen — CHANGES_REQUESTED** [review]

File: src/ContentCreator.php (hunk 1 above)
Comment (mtodor) [reply src/ContentCreator.php]: 👍

File: src/ContentCreator.php (hunk 2 above)
Comment (mtodor) [reply src/ContentCreator.php]: The conclusion from the chat:
- in the content creator, we will skip the creation of reference if target bundles are empty
- in distribution, we will add `media_entity_generic` in built artifact
- in Thunder performance docker we will enable `media_entity_generic`

Comment (mtodor) [reply src/ContentCreator.php]: - the skip of reference creation is added to this PR
- PR for distribution is provided here: https://github.com/thunder/thunder-distribution/pull/67
- and enabling of module for Thunder performance docker is added in this PR: https://github.com/thunder/docker-thunder-performance/pull/2

**Comment by mtodor** [comment]
@dbosen I think all your comments are addressed

**Review by dbosen — APPROVED** [review]
This works for me now

**Pull request merged**